CharlieBackend.Business/Services/AccountService.cs

```diff
@@ -170,5 +171,61 @@ public async Task<Result<AccountDto>> ChangePasswordAsync(ChangeCurrentPasswordD
 
             return Result<AccountDto>.GetError(ErrorCode.InternalServerError, "Salt for this account does not exist.");
         }
+
+        public async Task SendChangePasswordUrlAsync(string formUrl, ForgotPasswordDto forgotPassword)
+        {
+            var user = await _unitOfWork.AccountRepository.GetAccountCredentialsByEmailAsync(forgotPassword.Email);
+
+            user.ForgotPasswordToken = Guid.NewGuid().ToString();
+            user.ForgotTokenGenDate = DateTime.Now;
+
+            await _unitOfWork.CommitAsync();
+
+            string callbackUrl = formUrl + "token=" + user.ForgotPasswordToken;
+
+            await _notification.ForgotPasswordNotify(forgotPassword.Email, callbackUrl);
+        }
+
+        public async Task<Result<AccountDto>> ResetPasswordAsync(string guid, ResetPasswordDto resetPassword)
+        {
+            var user = await _unitOfWork.AccountRepository.GetAccountCredentialsByEmailAsync(resetPassword.Email);
+
+            if (user == null)
+            {
+                return Result<AccountDto>.GetError(ErrorCode.NotFound, "Account does not exist.");
+            }
+
+            if (user.ForgotPasswordToken != guid)
+            {
+                return Result<AccountDto>.GetError(ErrorCode.ValidationError, "Form validation error.");
+            }
+
+            if (!user.ForgotTokenGenDate.HasValue)
```

Comment:
join this check to line:210. If we do not have the date (which is quite low probability case), it would be ok to say that token is expired (therefore, new forgot password request will fix this)